arangod/Scheduler/SupervisedScheduler.cpp

```diff
@@ -249,20 +251,43 @@ SupervisedScheduler::SupervisedScheduler(application_features::ApplicationServer
                                arangodb_scheduler_medium_prio_queue_length{}),
                            _server.getFeature<arangodb::MetricsFeature>().add(
                                arangodb_scheduler_low_prio_queue_length{})} {
-  _queues[0].reserve(maxQueueSize);
-  _queues[1].reserve(fifo1Size);
-  _queues[2].reserve(fifo2Size);
-  _queues[3].reserve(fifo3Size);
+  _queues[0].queue.reserve(maxQueueSize);
+  _queues[1].queue.reserve(fifo1Size);
+  _queues[2].queue.reserve(fifo2Size);
+  _queues[3].queue.reserve(fifo3Size);
   TRI_ASSERT(fifo3Size > 0);
 }
 
 SupervisedScheduler::~SupervisedScheduler() = default;
 
-bool SupervisedScheduler::queueItem(RequestLane lane, std::unique_ptr<WorkItemBase> work) {
+bool SupervisedScheduler::queueItem(RequestLane lane, std::unique_ptr<WorkItemBase> work, bool bounded) {
   if (!_acceptingNewJobs.load(std::memory_order_relaxed)) {
     return false;
   }
 
+  auto const queueNo = static_cast<size_t>(PriorityRequestLane(lane));
+  TRI_ASSERT(queueNo < NumberOfQueues);
+
+  auto& queue = _queues[queueNo];
+  if (bounded) {
+    auto maxSize = _maxFifoSizes[queueNo];
+    auto queueFull = [&](){
+      LOG_TOPIC("98d94", DEBUG, Logger::THREADS)
+        << "unable to push job to scheduler queue: queue is full";
+      logQueueFullEveryNowAndThen(queueNo, maxSize);
+      ++_metricsQueueFull;
+    };
+    if (queue.numCountedItems.load(std::memory_order_relaxed) > maxSize) {
```

Comment:
What do we achieve with this load before the `fetch_add`? True, if the queue is full, we save two stores. But we pay in the non-full case with an additional load. I would probably have optimized for the queue not full case and paid with two stores if the queue is full...

Reply:
I agree. We should unconditionally execute `fetch_add(1, ...)` without doing an extra check before. That way we may over-count the number of queued items and may even drop a few requests. But this will only happen when we are close to the bound anyway, so I don't think this is a problem if we treat the bounds as "best effort" exact only.

Reply:
Fair enough, will change.